Sources/SwiftGRPCNIO/ClientCalls/ClientCall.swift

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2019, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import Foundation
+import NIO
+import NIOHTTP1
+import NIOHTTP2
+import SwiftProtobuf
+
+/// Base protocol for a client call to a gRPC service.
+public protocol ClientCall {
+  /// The type of the request message for the call.
+  associatedtype RequestMessage: Message
+  /// The type of the response message for the call.
+  associatedtype ResponseMessage: Message
+
+  /// HTTP/2 stream that requests and responses are sent and received on.
+  var subchannel: EventLoopFuture<Channel> { get }
+
+  /// Initial response metadata.
+  var initialMetadata: EventLoopFuture<HTTPHeaders> { get }
+
+  /// Status of this call which may be populated by the server or client.
+  ///
+  /// The client may populate the status if, for example, it was not possible to connect to the service.
+  ///
+  /// Note: despite `GRPCStatus` being an `Error`, the value will be __always__ delivered as a __success__
+  /// result even if the status represents a __negative__ outcome. This future will __never__ be fulfilled
+  /// with an error.
+  var status: EventLoopFuture<GRPCStatus> { get }
+
+  /// Trailing response metadata.
+  ///
+  /// This is the same metadata as `GRPCStatus.trailingMetadata` returned by `status`.
+  var trailingMetadata: EventLoopFuture<HTTPHeaders> { get }
+
+  /// Cancel the current call.
+  ///
+  /// Closes the HTTP/2 stream once it becomes available. Additional writes to the channel will be ignored.
+  /// Any unfulfilled promises will be failed with a cancelled status (excepting `status` which will be
+  /// succeeded, if not already succeeded).
+  func cancel()
+}
+
+extension ClientCall {
+  public var trailingMetadata: EventLoopFuture<HTTPHeaders> {
+    return status.map { $0.trailingMetadata }
+  }
+}
+
+/// A `ClientCall` with request streaming; i.e. server-streaming and bidirectional-streaming.
```

Comment:
"i.e. client-streaming"

Reply:
Sorry, _still_ doesn't look resolved to me.

Reply:
Haven't pushed this yet! I marked the trivial things as resolved that I've sorted locally.